### DISCUSSION TITLE
Master code set and get mechanism in doorlock through UI

### DISCUSSION THREAD

**Discussion by priyanka043** [zwave-js/zwave-js#6527] [Q&A] [answered]
Hi,

Is it possible to set and get the master code of the doorlock through UI

**Comment by robertsLando** [comment]
Sure, just go to node, user code CC, there you can find the master code input to get/set it

**Comment by priyanka043** [comment]
But here, i'm only able to see the set/get for user codes 1-250 not for the master code, can you please check the attached file
![Uploading Screenshot from 2023-07-25 10-56-55.png…]()

**Reply by robertsLando** [reply]
@priyanka043 I can't see the screenshot. You should wait it to end uploading before submitting the reply

**Comment by priyanka043** [comment]
![Uploading Screenshot from 2023-11-15 17-54-04.png…]()

**Reply by robertsLando** [reply]
Same as above, screenshot not visible

**Comment by priyanka043** [comment]
![Screenshot from 2023-11-15 17-53-51](https://github.com/zwave-js/zwave-js-ui/assets/139616028/ca92eb9d-bdf8-41d2-9141-eb9dd3c7e4c2)
![Screenshot from 2023-11-15 17-54-04](https://github.com/zwave-js/zwave-js-ui/assets/139616028/c8ab1344-78c5-4c73-a464-f1662b13bdd4)


PFA logs

**Reply by robertsLando** [reply]
Go on node debug info tab and paste here the content

**Comment by priyanka043** [comment]
Hii Daniel,

Here I am attaching the debug info data for the Yale Lock

**Reply by robertsLando** [reply]
I don't see it

**Comment by priyanka043** [comment]
HI

PFA log

**Reply by robertsLando** [reply]
Still don't see it. Notice that attached files to emails are not attached here on Github

**Comment by priyanka043** [comment]
[doorlock debuginfo.pdf](https://github.com/zwave-js/zwave-js-ui/files/13425605/doorlock.debuginfo.pdf)
Hi, please check this log file, please guide us to set / get mastercode for doorlock

**Reply by robertsLando** [reply]
Seems master code is not present on values. @priyanka043 are you sure that device supports master code?

cc @AlCalzone

**Reply by AlCalzone** [reply]
Please make a [driver log](https://zwave-js.github.io/zwave-js-ui/#/troubleshooting/generating-logs?id=driver-logs) of a re-interview, loglevel `debug` and attach it here as a file (drag & drop into the text field).

**Comment by priyanka043** [comment]
[www-out.log](https://github.com/zwave-js/zwave-js-ui/files/13501193/www-out.log)
can you please check this log while i was adding the lock

**Reply by AlCalzone** [reply]
That's not a driver log. Read the link from my last post again.
Also you don't have to re-add it. Re-interviewing is enough.

**Comment by priyanka043** [comment]
can you please let me know where can i get the driver log. like how can i take that

**Reply by AlCalzone** [reply]
![grafik](https://github.com/zwave-js/zwave-js-ui/assets/17641229/89c35c7a-3dee-42b0-ab4d-2165e3e1fc85)

**Comment by priyanka043** [comment]
can yo please check this log
[zwavejs_2023-11-29.log](https://github.com/zwave-js/zwave-js-ui/files/13501574/zwavejs_2023-11-29.log)

**Reply by AlCalzone** [reply]
You need to re-interview the device **while** making the log.
![grafik](https://github.com/zwave-js/zwave-js-ui/assets/17641229/73264642-303b-4446-b384-a4b1f7e4bf25)

Then wait until the status shows Complete again.

**Comment by priyanka043** [comment]
please check this
[zwavejs_2023-11-29.log](https://github.com/zwave-js/zwave-js-ui/files/13501911/zwavejs_2023-11-29.log)

**Comment by AlCalzone** [comment, marked as answer]
Your device reports support for User Code CC, version 1. Master codes are only available in version 2 of that CC.

**Comment by priyanka043** [comment]
can you please let me know where can we confirm that the doorlock is version 2 from the log.

**Reply by AlCalzone** [reply]
There are lines like
```
CNTRLR » [Node 013]   querying the CC version for Door Lock...
```
and then a bit below:
```
[Node 013]   supports CC Door Lock (0x62) in version 2
```

You'll have to distinguish between Door Lock CC and User Code CC. Door Lock CC is concerned with locking/unlocking the door.
User Code CC is concerned with managing the PIN codes. This one needs to be version 2 for master code support, but it is version 1:
```
2023-11-29T14:43:59.167Z DRIVER « [Node 013] [REQ] [BridgeApplicationCommand]
                                  │ RSSI: -89 dBm
                                  └─[VersionCCCommandClassReport]
                                      CC:      User Code
                                      version: 1
2023-11-29T14:43:59.169Z CNTRLR   [Node 013]   supports CC User Code (0x63) in version 1
```

**Comment by priyanka043** [comment]
thanks for your help, i have one more doubt, like when device user code is deleted the schedules related to that particular user code should automatically delete right?

**Comment by priyanka043** [comment]
![yale_version](https://github.com/zwave-js/zwave-js-ui/assets/139616028/67f92ceb-56cf-4e81-ab7d-2289d8c2af24)
Here, it is shown in yale manual that version 1 can be accessed through read/write , do you have any idea on this?

**Reply by AlCalzone** [reply]
Ok that looks like a workaround for the missing Master Code support in version 1.
0xFB is slot 251, which is just like any other user code from the Z-Wave perspective.

But... the lock reports that it only supports 250 codes, and Z-Wave JS prevents you from accessing other slots.
We'd have to override the response from the lock and hope that this enables you to access that slot.

**Reply by robertsLando** [reply]
Dunno if maybe they mean that settings one of the last user codes equals to set the master code. BTW that is so weird

**Comment by AlCalzone** [comment]
@priyanka043 I'm transferring this discussion to the `node-zwave-js` repo for further investigation.

**Comment by AlCalzone** [comment]
@priyanka043 let's try this:

- [This file](https://raw.githubusercontent.com/zwave-js/node-zwave-js/6a1a79589ef18ca2e2b4fac758c86b4d201bdc30/packages/config/config/devices/0x0129/yrc_yrd_226_246_256_446-zw2.json) has an override for the slot count, which should allow you to access slot 251 (the master code)
- In Z-Wave JS UI in the store tab there is a folder named `config`:
    ![grafik](https://github.com/zwave-js/node-zwave-js/assets/17641229/b543244c-25c0-4b1c-9141-997aacdacb2a)
    Download the above file and put it there.
- Restart Z-Wave JS UI
- Re-interview your device

You should now have 251 lock code slots where number 251 is your master code. You'll need to refresh the lock codes manually with the [REFRESH] button, as this isn't done automatically during the interview.

Let me know if that works, so I can roll out the change for everyone.

**Comment by priyanka043** [comment]
How to put it there, actually there is a folder config there i can see only
either delete or add the file name

**Comment by priyanka043** [comment]
[yrc_yrd_226_246_256_446-zw2.json](https://github.com/zwave-js/node-zwave-js/files/13513696/yrc_yrd_226_246_256_446-zw2.json)
![Screenshot from 2023-11-30 19-31-01](https://github.com/zwave-js/node-zwave-js/assets/139616028/1bec9484-3a10-4732-a6ff-15c7e2793511)
The above is the json file i have uploaded, you can see with the last name

**Comment by priyanka043** [comment]
I have restarted zwavejs ui, and re-interviewed the node, refreshed usercode CC but still able to see only 250 keys

**Comment by priyanka043** [comment]
Did i put the file at the right place? like i have selected config and i have uploaded the file

**Reply by AlCalzone** [reply]
No, it needs to be in the config directory. It looks like it is next to it. @robertsLando ?

**Reply by robertsLando** [reply]
Yeah the location is wrong, that's not the config directory

**Comment by priyanka043** [comment]
So, in config directory do I need to create any other directory and place
this file or how?

**Reply by AlCalzone** [reply]
No just put the file **inside** the config directory

**Reply by robertsLando** [reply]
@priyanka043 Don't you have direct access to that folder? If so use thst way instead of the ui...

Using the ui should be easy, next to every folder (on the right) there is a button to create a new file, once you create the file give it the name and then paste inside the content. You cannot directly import a file in a folder or move them around AFAIK

**Comment by priyanka043** [comment]
if I select the config then it's asking for the file name but unable to
update the file in that

**Comment by priyanka043** [comment]
Hi,

I have tried to create the file and i have copied the data, even though im
not able to see anything regarding slot 251

PFA image.

**Comment by priyanka043** [comment]
![Screenshot from 2023-12-04 19-25-25](https://github.com/zwave-js/node-zwave-js/assets/139616028/28fc1652-58d4-45db-a1f6-9a24fda6806d)

**Comment by priyanka043** [comment]
Is that file is placed at the correct location?

**Comment by AlCalzone** [comment]
Looks good. Have you restarted Z-Wave JS UI and re-interviewed the device yet?
I'd like to see a driver log (level debug) of that re-interview to confirm the change does what it is supposed to.

**Comment by priyanka043** [comment]
[zwavejs_2023-12-08.log](https://github.com/zwave-js/node-zwave-js/files/13613395/zwavejs_2023-12-08.log)
PFA driver log after re-interview node

**Comment by AlCalzone** [comment]
```
2023-12-08T11:26:46.098Z DRIVER   all queues idle
2023-12-08T11:26:46.206Z CNTRLR   [Node 001] Embedded device config loaded
2023-12-08T11:26:46.518Z CNTRLR   [Node 013] Embedded device config loaded
2023-12-08T11:26:46.708Z CNTRLR   [Node 016] Embedded device config loaded
2023-12-08T11:26:46.749Z CNTRLR   [Node 019] No device config found
2023-12-08T11:26:46.908Z CNTRLR   [Node 028] Embedded device config loaded
2023-12-08T11:26:46.981Z CNTRLR   [Node 037] Embedded device config loaded
2023-12-08T11:26:47.052Z CNTRLR   [Node 038] Embedded device config loaded
2023-12-08T11:26:47.082Z CNTRLR   [Node 050] Embedded device config loaded
2023-12-08T11:26:47.134Z CNTRLR   [Node 058] Embedded device config loaded
2023-12-08T11:26:47.166Z CNTRLR   [Node 060] Embedded device config loaded
2023-12-08T11:26:47.279Z CNTRLR   [Node 063] Embedded device config loaded
2023-12-08T11:26:47.324Z CNTRLR   [Node 070] Embedded device config loaded
2023-12-08T11:26:47.420Z CNTRLR   [Node 083] Embedded device config loaded
2023-12-08T11:26:47.555Z CNTRLR   [Node 085] Embedded device config loaded
2023-12-08T11:26:47.580Z CNTRLR   [Node 086] Embedded device config loaded
2023-12-08T11:26:47.591Z CNTRLR   [Node 087] No device config found
2023-12-08T11:26:47.643Z CNTRLR   [Node 088] Embedded device config loaded
2023-12-08T11:26:47.776Z CNTRLR   [Node 089] Embedded device config loaded
2023-12-08T11:26:47.826Z CNTRLR   [Node 093] Embedded device config loaded
2023-12-08T11:26:47.951Z CNTRLR   [Node 094] Embedded device config loaded
2023-12-08T11:26:47.962Z CNTRLR   [Node 098] No device config found
2023-12-08T11:26:48.019Z CNTRLR   [Node 109] Embedded device config loaded
2023-12-08T11:26:48.066Z CNTRLR   [Node 110] Embedded device config loaded
2023-12-08T11:26:48.101Z CNTRLR   [Node 111] Embedded device config loaded
```
Doesn't look like the custom config file was picked up.

Also, I need to see a log of the re-interview, not afterwards.

**Comment by priyanka043** [comment]
Hi PFA logs 
[zwavejs_current.log](https://github.com/zwave-js/node-zwave-js/files/13631151/zwavejs_current.log)
[zwavejs_2023-12-09.log](https://github.com/zwave-js/node-zwave-js/files/13631152/zwavejs_2023-12-09.log)

**Reply by priyanka043** [reply]
Please help us to get and set the master code provisioning. if any thing logs needed, let me know i will be updating.

**Comment by priyanka043** [comment]
Hi, 

Did you see any mechanism to do this

**Comment by AlCalzone** [comment]
I'll answer when I find some time --> https://github.com/zwave-js/node-zwave-js/issues/6521

**Comment by priyanka043** [comment]
Hi, 

How it is possible to send the raw data to zwavejs, i mean the zwave raw commands to controller, can you please provide any example to send any raw data to on/off a switch.

**Reply by robertsLando** [reply]
This is possible using driver function from ui with a custom snippet

**Comment by priyanka043** [comment]
Thanks,

I have tried to add the door loack with PC controller and i tried to get the usercode of 251, here i can get the master code, i have captured this through pc ziffer, i would be a great help if it possible to get/ set the master code with zwavejsui, PFA 
![image](https://github.com/zwave-js/node-zwave-js/assets/139616028/05b37f4c-4ace-466c-86aa-9204701b4171)

**Comment by priyanka043** [comment]
Hi, 

Can i get a driver function code to set a user code of a door lock?

**Comment by priyanka043** [comment]
As we can see in the source code if we want to set the code for user code from1 -250, i'm able to set but i'm unable to set the code to 251, i think its due to restrictions in source code. that it is allowing only 1-250, can you please help me to set the user code 251. i.e, the master code of the door lock

**Comment by AlCalzone** [comment]
First off: I'm on [parental leave](https://github.com/zwave-js/node-zwave-js/issues/6521), so I may be slow to respond.

It appears that the lock you interviewed in the last log you posted has an entirely different ID than the one I edited the config file for:
> OLD:
manufacturer id: 0x0129
product type:    0x8002
product id:      0x0600

> NEW:
manufacturer id: 0x0129
product type:    0xffff
product id:      0x190f

Not sure if this is a mistake, or the device reports something else, or you interviewed a different lock? The new one doesn't even have a config file in our database.

Assuming this still is the same device, I've added the fingerprint to the modified file I sent you before:
[yrc_yrd_226_246_256_446-zw2.json](https://github.com/zwave-js/node-zwave-js/files/13727103/yrc_yrd_226_246_256_446-zw2.json)
Please replace the file in the `config` directory with this new one, then restart Z-Wave JS UI again, and re-interview your device again.
Ideally make another logfile, so I have something to investigate when this doesn't work.

**Reply by AlCalzone** [reply]
@priyanka043 did this work?

**Reply by priyanka043** [reply]
yes, I have changed this in server/node_modules/@zwave-js/config/config/devices path and then its working

**Reply by AlCalzone** [reply]
Could you make another log of a re-interview please, so I have all the info I need to make this fix available for everyone?

**Reply by AlCalzone** [reply]
@priyanka043 reminder :)
